### PR TITLE
ACS-306: ACS Repository: V1 Rest API for content direct access urls

### PR DESCRIPTION
**Pull request by cristianturlica**
- added alfresco.content.directAccessUrl.expiryTime.seconds property to allow the configuration of the direct access url expiry time
   - changed getDirectAccessUrl method signature to allow the successful execution of the acl interceptor
   - moved get content data to it's own method so it can be easily reused
   - updated test

### CODE REVIEW

File: src/main/resources/alfresco/repository.properties [hunk 1]
```diff
@@ -1327,4 +1327,8 @@ system.delete_not_exists.batchsize=100000
 system.delete_not_exists.delete_batchsize=1000
 system.delete_not_exists.read_only=false
 system.delete_not_exists.timeout_seconds=-1
-system.prop_table_cleaner.algorithm=V2
\ No newline at end of file
+system.prop_table_cleaner.algorithm=V2
+
+# Configure the expiration time of the direct access url.
+# Note: It is up to the actual ContentStore implementation if it can fulfil this request or not.
+alfresco.content.directAccessUrl.expiryTime.seconds=-1
```
Comment (ancutam) [inline src/main/resources/alfresco/repository.properties:1334]: What does `-1` mean? No expiry time?

File: src/main/java/org/alfresco/service/cmr/repository/ContentService.java [hunk 2]
```diff
@@ -157,9 +157,10 @@ public ContentWriter getWriter(NodeRef nodeRef, QName propertyQName, boolean upd
      * Gets a presigned URL to directly access a binary content. It is up to the content store
      * if it can fulfil this request with an expiry time or not.
      *
-     * @param contentUrl A content store URL
+     * @param nodeRef a reference to a node having a content property
      * @param expiryTime Expiration time in milliseconds
```
Comment (gavincornwell) [inline src/main/java/org/alfresco/service/cmr/repository/ContentService.java:161]: I think we need to be a little clearer with the explanation of the expiryTime parameter. Is it the time in epoch seconds when the link expires or the length of time in milliseconds that the link is valid for?

File: src/main/resources/alfresco/repository.properties (hunk 1 above)
Comment (gavincornwell) [reply src/main/resources/alfresco/repository.properties:1334]: I don't think it should be possible to get a link that never expires (not even sure the AWS API allows for that) so we should have a reasonable default i.e. 24 hours?

File: src/main/java/org/alfresco/service/cmr/repository/ContentService.java (hunk 2 above)
Comment (cristianturlica) [reply src/main/java/org/alfresco/service/cmr/repository/ContentService.java:161]: The length of time in milliseconds that the link is valid for. I will update the description.

File: src/main/resources/alfresco/repository.properties (hunk 1 above)
Comment (cristianturlica) [reply src/main/resources/alfresco/repository.properties:1334]: The initial intent was that -1 - would imply not set and allowed for each connecter default to be used. But I can change it to 24 hours.

Comment (cristianturlica) [reply src/main/resources/alfresco/repository.properties:1334]: Amazon sdk has something similar and defaults to 15 minutes if not set...

Comment (cristianturlica) [reply src/main/resources/alfresco/repository.properties]: Set to 86400 seconds (24 hours)

File: src/main/java/org/alfresco/service/cmr/repository/ContentService.java [hunk 3]
```diff
@@ -154,12 +154,18 @@ public ContentWriter getWriter(NodeRef nodeRef, QName propertyQName, boolean upd
...
-     * @param contentUrl A content store URL
-     * @param expiryTime Expiration time in milliseconds 
-     * @return A direct access URL for a binary content or empty string if not supported
+     * @param nodeRef
+     *            a reference to a node having a content property
+     * @param expiryTime
+     *            the expiration time in milliseconds of the direct access url. This
+     *            is the length of time in milliseconds that the link is valid for.
+     * @return A direct access URL for a binary content or empty string if not
+     *         supported
```
Comment (ancutam) [inline src/main/java/org/alfresco/service/cmr/repository/ContentService.java:167]: or returns `null` if there is no binary content for the node

Comment (DenisGabriela) [reply src/main/java/org/alfresco/service/cmr/repository/ContentService.java:167]: The AWS API generates an URL even though there is no binary content on the given path.

Comment (ancutam) [reply src/main/java/org/alfresco/service/cmr/repository/ContentService.java:167]: Our ContentService returns null though.

Comment (DenisGabriela) [reply src/main/java/org/alfresco/service/cmr/repository/ContentService.java:167]: You're right. I was thinking of binary content in s3.

File: src/main/resources/alfresco/repository.properties (hunk 1 above)
Comment (cristianturlica) [reply src/main/resources/alfresco/repository.properties]: Updated to 28800 (8 hours) - after talking with Alex Lee

File: src/main/java/org/alfresco/service/cmr/repository/ContentService.java (hunk 3 above)
Comment (cristianturlica) [reply src/main/java/org/alfresco/service/cmr/repository/ContentService.java]: Change done, good catch.

**Pull request merged**